### ISSUE TITLE
Add a Navigation Library

### ISSUE THREAD

**Issue by orta** [artsy/emission#501]
For consignments I need to show a modal, with navigation between VCs inside it. I could push this to Eigen, but realistically this is something that over time we'll want to do more of inside Emission.

Current choices:

- [ ] NavigatorIOS - https://facebook.github.io/react-native/docs/navigatorios.html
- [ ] Navigation https://facebook.github.io/react-native/docs/navigation.html
- [ ] https://github.com/expo/ex-navigation
- [ ] https://github.com/wix/react-native-navigation 
- [ ] http://airbnb.io/native-navigation/ - 👎 Swift

Will edit this issue as I look at each

**Comment by alloy** [comment]
Yeah all non-native options are an absolute no go imo.

Bummer that Native Navigation is done in Swift, I really don’t want any Swift in there. Does NavigatorIOS support callbacks from pushed/modal VCs, [like Native Navigation does](http://airbnb.io/native-navigation/docs/api/navigator/push.html#returns)?

**Comment by orta** [comment]
Hrm, I'm not sure what you mean by callbacks - you mean the `NavigationResult` which is returned? 

I'm guessing the answer is no FWIW. I think [React-Native Navigation](https://wix.github.io/react-native-navigation/#/top-level-api) has the same kind of API.

**Comment by alloy** [comment]
Yes, the `Promise<NavigationResult>` return value. It seems very handy, in retrospect, for e.g. the refine view.

**Comment by orta** [comment]
Agreed, [present](http://airbnb.io/native-navigation/docs/api/navigator/present.html)'s API is good here: 

> (Promise<NavigationResult>): A promise that resolves when the presented screen gets dismissed.

Not useful enough to un-warrent the Swift drawback, and it doesn't look like there's something [_directly_](https://wix.github.io/react-native-navigation/#/screen-api?id=pushparams) [like](https://wix.github.io/react-native-navigation/#/screen-api?id=showmodalparams-) this in React Native Navigation

**Comment by alloy** [comment]
Oh no, totally, def not adding a Swift dependency.

A quick look at the code reveals that the code is probably too large to do an easy port to objc and I also doubt that @lelandrichardson would accept a PR for that (not technically required of course, but means we need to maintain a larger than a few LOC dependency).

So maybe we should just keep our own lean version that does just what we need and steal some ideas from Native Navigation?

**Comment by KrauseFx** [comment]
I'm actually very interested in the outcome of this, I had the same question when deciding what to use for the [wwdc.family](https://github.com/wwdc-family/app/) app, and was surprised that there is no good native built-in ones that support both platforms.

**Comment by alloy** [comment]
@KrauseFx If a Swift dependency is ok for you, I would definitely suggest you checkout Native Navigation. The talk about it from React conf makes it look pretty sweet https://www.youtube.com/watch?v=tWitQoPgs8w

**Comment by KrauseFx** [comment]
Yeah, I don't really feel like adding 15MB to my binary :trollface:

**Comment by alloy** [comment]
Right, if you don't have Swift libs in there yet it's an especially tough sell.

**Comment by maicki** [comment]
I looked into this a while back too and as I had the same problems with all the other solutions (it's especially unfortunate that native navigation is in Swift) so I implemented a pretty rough small homegrown solution in objc: https://github.com/maicki/ReactNativeNavigationPlayingAround just to get an idea how this could work. As said it's pretty rough, but the idea was just to get some ideas running without really spend so much time on file separation etc.

There are actually a couple of different ways how you could tackle that, either you can go the view controller way that receives events from the RN navigation components or have a root view that listens for navigation events. I build both in there with examples how to use a `ReactNativeNavigationController` and how to just use a plain `UIViewController` that has a specific `RCTRootView` subclass (`ReactNativeNavigation`) as root view that listens to navigation events and forwards it to a delegate.

@alloy @orta maybe this is helpful if you decide to build your own lean version.

@KrauseFx By coincident I got an invite today for wwdc.family so I could install the app on my phone. I think it would not be too difficult to build your screens on top of a regular view controller stack, but it also depends how you bootstrapped the RN project, e.g. if you used `react-native init` or created the app from scratch via Xcode.

**Comment by orta** [comment]
@maicki - this is definitely enough for us to get started with. I ran through the demo, and it has all the sorts of things we would need _today_. I'll probably look into adding all these in the next two weeks - so thanks for the option, I may turn it into something 👍

**Comment by alloy** [comment]
@maicki Nice, thanks!

**Comment by orta** [comment]
From inside the React-Native Navigation discord:

![screen shot 2017-05-08 at 11 03 08](https://cloud.githubusercontent.com/assets/49038/25799702/f854f8ec-33dd-11e7-8a0b-5df6b294e95b.png)

**Comment by DanielZlotin** [comment]
I would like to step in and say that features like returning a Promise from commands (that will resolve once the transition is complete) is in the roadmap of react-native-navigation.

I'm happy for any and all feedback, let's make navigation great again 😄

**Comment by tamastimar** [comment]
@orta Hi! Coming from [Intro to React Native for an iOS Developer](http://artsy.github.io/blog/2017/07/06/React-Native-for-iOS-devs/).

> ## [Navigation](https://facebook.github.io/react-native/docs/navigation.html) **_Official_**
> Production worthy and in RN core, but is basically React Navigation below.

That link is pointing to an overview article where there is no component mentioned which is called Navigation.

Is that a thing still available?

**Comment by orta** [comment]
Ah, I'll re-write that, I think I was going for "this is the recommended pattern, but it's basicall react-navigation"

**Comment by orta** [comment]
Check-out - https://github.com/APSL/react-native-navigator-wrapper

**Comment by aksswami** [comment]
@orta Above navigation wrapper is deprecated. So what is the current approach from navigation for emission?

**Comment by alloy** [comment]
@aksswami Still the same tiny homegrown solution we’ve been using from the start. Search this repo for all files related to ‘switchboard’.

**Comment by orta** [comment]
https://github.com/maicki/react-native-router-bridge - this can bridge our existing `ARSwitchBoard` with react-router

**Comment by brentvatne** [comment]
I don't think this will impact your decision at all and it's not my intention to attempt to, I just wanted to point out for posterity that the original screenshots that show every stack screen in the view hierarchy is an issue that has been (experimentally) solved in react-navigation with https://github.com/kmagiera/react-native-screens.

**Comment by orta** [comment]
Awesome, I'll edit the original post to include this 👍
